jax/_src/lax/control_flow/loops.py

```diff
@@ -1254,87 +1218,58 @@ def _scan_partial_eval_custom(saveable, unks_in, inst_in, eqn):
     assert False, "Fixpoint not reached"
   jaxpr_known  = core.ClosedJaxpr(jaxpr_known_ , jaxpr.consts)
   jaxpr_staged = core.ClosedJaxpr(jaxpr_staged_, jaxpr.consts)
-
-  # Move all residual binders to the back of jaxpr_staged so they're extensive.
-  # TODO(mattjj): make jaxpr_staged only take instantiated inputs
   res_avals = jaxpr_staged.in_avals[:num_res]
-  jaxpr_staged = pe.move_binders_to_back(
-      jaxpr_staged, [True] * num_res + [False] * len(jaxpr.in_avals))
 
   # Instantiate all inputs (b/c jaxpr_staged takes all inputs, corresponding to
   # passing in_inst argument to partial_eval_jaxpr_custom above).
   new_inst = [x for x, inst in zip(eqn.invars, inst_in)
               if type(x) is core.Var and not inst]
   inst_in = [True] * len(inst_in)
 
-  # As an optimization, hoist loop-invariant residuals out of the loop rather
-  # than using extensive outputs for them. See _scan_partial_eval for comments.
-  num_const_known = len(const_uk) - sum(const_uk)
-  num_carry_known = len(carry_uk) - sum(carry_uk)
-  num_xs_known    = len(   xs_uk) - sum(   xs_uk)
-  jaxpr_known_hoist, jaxpr_known_loop, loop_dep, consts_known_lp_avals = \
-      pe.partial_eval_jaxpr_nounits(
-          jaxpr_known,
-          [False] * num_const_known + [True] * (num_carry_known + num_xs_known),
-          [True] * (len(unks_out) - sum(unks_out)) + [False] * num_res)
-  # jaxpr_known_hoist produces intensive residuals followed by the constants for
-  # jaxpr_known_loop. We adjust jaxpr_staged to accept intensive res as consts.
-  _, loop_dep_res = split_list(loop_dep, [len(loop_dep) - num_res])
-  jaxpr_staged = pe.move_binders_to_front(
-      jaxpr_staged, [False] * sum(inst_in) + _map(operator.not_, loop_dep_res))
-  num_intensive_res = len(loop_dep_res) - sum(loop_dep_res)
-  del loop_dep, num_carry_known, num_xs_known, const_uk
-
-  # Create residual variables.
-  intensive_avals, ext_avals_mapped = partition_list(loop_dep_res, res_avals)
-  ext_avals = [core.unmapped_aval(eqn.params['length'], 0, a)
-               for a in ext_avals_mapped]
+  # No res forwards should be possible because jaxpr_unknown already has all
+  # inputs instantiated.
+  num_knowns_out = len(jaxpr_known.out_avals) - num_res
+  num_consts_known = len(const_uk) - sum(const_uk)
+  num_carry_known  = len(carry_uk) - sum(carry_uk)
+  in_fwd: list[int | None] = pe._jaxpr_forwarding(jaxpr_known.jaxpr)
+  in_fwd = [f if out_idx >= num_knowns_out and f is not None and
+            (f < num_consts_known or f >= num_consts_known + num_carry_known)
+            else None for out_idx, f in enumerate(in_fwd)]
+  assert all(f is None for f in in_fwd)
+  del in_fwd
```

Comment:
We could remove this check, I guess? Although perhaps it's not too costly if jaxpr_forwarding is cached...

Reply:
I agree this seems costly. I put it in here as a sanity check, but we should probably remove it... thanks for catching it.